reporter/jira.go

```diff
@@ -0,0 +1,197 @@
+			return newJiraReporter(*jiraURL, *username, *password, *fieldsConfiguration, *closedStatus)
+		},
+	})
+
+}
+
+func newJiraReporter(jiraURL, username, password, fieldsConfiguration, closedStatus string) (*jiraReporter, error) {
+
+	err := checkArgsNotNil(jiraURL, username, password, fieldsConfiguration, closedStatus)
+	if err != nil {
+		return nil, err
+	}
+	reporter := new(jiraReporter)
+	jiraClient, err := jira.NewClient(nil, jiraURL)
+	if err != nil {
+		return nil, fmt.Errorf("Could not create client: %s", err)
+
+	}
+
+	res, err := jiraClient.Authentication.AcquireSessionCookie(username, password)
+	if err != nil || !res {
+		return nil, fmt.Errorf("Authentication failed: %s", err)
+	}
+
+	if !jiraClient.Authentication.Authenticated() {
+		return nil, fmt.Errorf("Not authenticated")
+	}
+
+	reporter.client = jiraClient
+
+	err = reporter.setFieldsConfig(fieldsConfiguration)
+	if err != nil {
+		return nil, err
+	}
+
+	project, found := reporter.fieldsConfig["Project"]
+	issueType, found := reporter.fieldsConfig["Issue Type"]
+
+	if !found {
+		return nil, fmt.Errorf("Project and Issue Type are required in field configuration")
+	}
+
+	reporter.closedStatusName = closedStatus
+
+	// get create meta information
+	meta, _, err := jiraClient.Issue.GetCreateMeta(project)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get create meta : %s", err)
+	}
+
+	// get right project
+	metaProject := meta.GetProjectWithKey(project)
+	if metaProject == nil {
+		return nil, fmt.Errorf("Could not find project with key %s", project)
+	}
+	reporter.metaProject = metaProject
+	// get right issue within project
+	metaIssuetype := metaProject.GetIssueTypeWithName(issueType)
+	if metaIssuetype == nil {
+		return nil, fmt.Errorf("Could not find issuetype %s. Available are %#v", issueType, getAllIssueTypeNames(metaProject))
+	}
+
+	// check if the given fields completes the mandatory fields and all listed fields are available
+	complete, err := metaIssuetype.CheckCompleteAndAvailable(reporter.fieldsConfig)
+	if !complete {
+		return nil, err
+	}
+
+	reporter.metaIssuetype = metaIssuetype
+	return reporter, nil
+}
+
+func (j *jiraReporter) Report(failure complainer.Failure, config ConfigProvider, stdoutURL, stderrURL string) error {
+
+	renderedFields := make(map[string]string)
+	// render all values as they can be tempaltes
+	for field, templatedValue := range j.fieldsConfig {
+		rendered, err := fillTemplate(failure, config, stdoutURL, stderrURL, templatedValue)
+		if err != nil {
+			return fmt.Errorf("Rendering value of %s as tempalte failed: %s", field, err)
+		}
+		renderedFields[field] = rendered
+	}
+
+	// generate jql with exact match for summary, project and status
+	query := fmt.Sprintf("summary ~ \"\\\"%s\\\"\" AND project = %s AND status != %s", renderedFields["Summary"], renderedFields["Project"], j.closedStatusName)
+	results, resp, err := j.client.Issue.Search(query, nil)
+	if err != nil {
+		return fmt.Errorf(readJiraReponse(resp))
```

Comment:
go-jira tries to return a proper struct in results if the call is successful. In case of error, like status code not in 2xx range, it will send the raw response body back without reading from it. That is the reason, we only have to read response if there is an error.